File: backend/src/main/java/com/team01/backend/domain/user/service/AuthService.java
package com.team01.backend.domain.user.service;

import com.team01.backend.domain.user.dto.*;
import com.team01.backend.domain.user.entity.Role;
import com.team01.backend.domain.user.entity.User;
import com.team01.backend.domain.user.repository.UserRepository;
import com.team01.backend.global.security.JwtTokenProvider;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * [과제명: 권한 계층 구조가 적용된 JWT 인증 시스템]
 * 본 서비스는 사용자의 가입 시점에 관리자 여부를 판별하고,
 * 이에 따른 차등화된 권한(USER/ADMIN)을 부여하는 핵심 로직을 수행합니다.
 */
@Service
@RequiredArgsConstructor
@Transactional
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;
	private final MailService mailService;

    /**
     * [메서드: signUp]
     * 관리자 토큰 검증 로직을 포함하여 사용자를 등록합니다.
     */
    @Transactional
    public void signUp(SignUpRequest request) {
        // 1. 입력값 기본 검증
        validateInput(request.getEmail(), request.getPassword());
        
        // [신규] 탈퇴 회원 포함 중복 체크: 이미 우리 시스템을 거쳐 간 모든 데이터를 확인합니다.
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new IllegalArgumentException("이미 사용 중이거나 탈퇴 처리된 이메일입니다.");
        }

        // 2. 관리자 여부 확인 및 권한 할당
        Role role = Role.USER;
        if (request.isAdmin()) {
            // 기획된 관리자 비밀 토큰과 일치하는지 확인 (BaseInitData와 동일값 설정)
            if (!"user_admin-2026".equals(request.getAdminToken())) {
                throw new IllegalArgumentException("관리자 인증 토큰이 일치하지 않습니다.");
            }
            role = Role.ADMIN;
        }

        // 3. 사용자 엔티티 생성 및 저장
		// profileImage를 넘기지 않아도 엔티티의 @PrePersist가 기본값을 할당함
        User user = User.builder()
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .nickname(request.getNickname())
                .role(role) // 결정된 권한 주입
                .build();

        userRepository.save(user);
    }

    /**
     * [메서드: login]
     * 인증 성공 시 사용자의 Role 정보가 포함된 실제 JWT를 발행합니다.
     */
    @Transactional
    public TokenDto login(LoginRequest request) {
        validateInput(request.getEmail(), request.getPassword());

        // 1. 전체 사용자에서 이메일 조회 (탈퇴 여부와 관계없이 조회)
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new IllegalArgumentException("이메일 또는 비밀번호가 일치하지 않습니다."));

        // [핵심 추가] 탈퇴한 회원 식별: WITHDRAWN 상태인 경우 명확한 에러 응답을 보냅니다.
        if (user.getRole() == Role.WITHDRAWN) {
            throw new IllegalArgumentException("탈퇴한 회원입니다.");
        }

        // 2. 비밀번호 검증
        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new IllegalArgumentException("이메일 또는 비밀번호가 일치하지 않습니다.");
        }

        // 3. 이중 토큰 발행: 액세스 토큰(단기)과 리프레시 토큰(장기)을 동시에 생성합니다.
        String accessToken = jwtTokenProvider.createAccessToken(user.getEmail(), user.getRole().name());
        String refreshToken = jwtTokenProvider.createRefreshToken(user.getEmail());

        // 4. 지휘관 각인: 리프레시 토큰을 DB에 저장하여 추후 세션 관리에 활용합니다.
        user.updateRefreshToken(refreshToken);
        
        return new TokenDto(accessToken, refreshToken);
    }

    /**
     * [신규] 로그아웃 로직
     * 사용자의 리프레시 토큰을 무효화하여 모든 기기에서의 세션 갱신을 차단합니다.
     */
    @Transactional
    public void logout(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new EntityNotFoundException("사용자를 찾을 수 없습니다."));
        
        // DB의 리프레시 토큰을 제거하여 탈취된 토큰의 사용을 원천 봉쇄합니다.
        user.updateRefreshToken(null);
    }

    /**
     * [신규] 토큰 재발급 로직 (Reissue)
     * 리프레시 토큰을 검증하고 새로운 액세스 토큰을 발행하여 끊김 없는 경험을 제공합니다.
     */
    @Transactional
    public String reissue(String refreshToken) {
        // 1. 토큰 자체의 유효성 검증
        if (!jwtTokenProvider.validateToken(refreshToken)) {
            throw new RuntimeException("리프레시 토큰이 만료되었습니다. 다시 로그인하십시오.");
        }

        // 2. 토큰 내 정보로 사용자 식별
        User user = userRepository.findByEmail(jwtTokenProvider.getUserEmail(refreshToken))
                .orElseThrow(() -> new EntityNotFoundException("사용자를 찾을 수 없습니다."));

        // 3. DB에 저장된 토큰과 대조하여 부적절한 갱신 시도를 차단합니다.
        if (!refreshToken.equals(user.getRefreshToken())) {
            throw new RuntimeException("비정상적인 토큰 갱신 시도입니다.");
        }

        // 4. 새로운 액세스 토큰 발행
        return jwtTokenProvider.createAccessToken(user.getEmail(), user.getRole().name());
    }

    /**
     * [신규] 아이디 찾기
     * 탈퇴한 유령 회원을 제외하고 오직 활성 사용자 중에서만 닉네임으로 검색합니다.
     */
    @Transactional(readOnly = true)
    public String findId(FindIdRequest request) {
        User user = userRepository.findByNicknameAndRoleNot(request.getNickname(), Role.WITHDRAWN)
                .orElseThrow(() -> new EntityNotFoundException("해당 정보로 등록된 활성 사용자가 없습니다."));
        return user.getEmail();
    }

	// 비밀번호 재설정 로직
    @Transactional
    public void resetPassword(PasswordResetRequest request) {
        // [보안 필수] 인증 코드 검증 단계
        boolean isVerified = mailService.verifyCode(request.getEmail(), request.getVerificationCode());
        if (!isVerified) {
            throw new IllegalArgumentException("인증 코드가 일치하지 않거나 만료되었습니다.");
        }

        User user = userRepository.findByEmailAndRoleNot(request.getEmail(), Role.WITHDRAWN)
                .orElseThrow(() -> new EntityNotFoundException("사용자를 찾을 수 없거나 이미 탈퇴한 계정입니다."));
        
        user.updatePassword(passwordEncoder.encode(request.getNewPassword()));
    }

    /**
     * [신규] 회원 탈퇴 수행 (Soft Delete)
     * 역할을 WITHDRAWN으로 변경하고 모든 인증 정보를 파기하여 영구 격리합니다.
     */
    @Transactional
    public void withdraw(String email) {
        User user = userRepository.findByEmailAndRoleNot(email, Role.WITHDRAWN)
                .orElseThrow(() -> new EntityNotFoundException("활성화된 사용자 정보를 찾을 수 없습니다."));
        
        user.withdraw(); // Role 변경 및 리프레시 토큰 null 처리
    }

    private void validateInput(String email, String password) {
        if (email == null || !email.contains("@")) {
            throw new IllegalArgumentException("올바른 이메일 형식이 아닙니다.");
        }
    }
}

File: backend/src/main/java/com/team01/backend/domain/user/service/AuthService.kt
package com.team01.backend.domain.user.service

import com.team01.backend.domain.user.entity.Role
import com.team01.backend.domain.user.entity.User
import com.team01.backend.domain.user.repository.UserRepository
import com.team01.backend.global.security.JwtTokenProvider
import com.team01.backend.global.security.TokenReissueException
import jakarta.persistence.EntityNotFoundException
import org.springframework.security.authentication.BadCredentialsException
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

data class AuthToken(
    val accessToken: String,
    val refreshToken: String
)

@Service
@Transactional
class AuthService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder,
    private val jwtTokenProvider: JwtTokenProvider
) {

    /**
     * 회원 가입을 처리합니다.
     * 동일한 이메일이 존재하면 예외를 발생시킵니다.
     */
    fun signUp(email: String, password: String, nickname: String) {
        if (userRepository.existsByEmail(email)) {
            throw IllegalArgumentException("이미 사용 중인 이메일입니다.")
        }

        val user = User(
            email = email,
            password = passwordEncoder.encode(password) ?: throw IllegalStateException("비밀번호 인코딩에 실패했습니다."),
            nickname = nickname,
            role = Role.USER 
        )
        userRepository.save(user)
    }

    /**
     * 로그인을 처리하고 새로운 AccessToken과 RefreshToken을 반환합니다.
     * 탈퇴한 회원이거나 비밀번호가 불일치하면 예외를 던집니다.
     */
    fun login(email: String, password: String): AuthToken {
        // Optional 제거 후 코틀린의 엘비스 연산자로 흐름을 완벽히 통제하네.
        val user = userRepository.findByEmail(email)
            ?: throw BadCredentialsException("이메일 또는 비밀번호가 일치하지 않습니다.")

        if (user.role == Role.WITHDRAWN) throw IllegalArgumentException("탈퇴한 회원입니다.")
        
        if (!passwordEncoder.matches(password, user.password)) {
            throw BadCredentialsException("이메일 또는 비밀번호가 일치하지 않습니다.")
        }

        val accessToken = jwtTokenProvider.createAccessToken(user.email, user.role.name)
        val refreshToken = jwtTokenProvider.createRefreshToken(user.email)

        // 영속성 엔티티의 리프레시 토큰 수정 (더티 체킹 발생)
        user.updateRefreshToken(refreshToken)

        return AuthToken(accessToken, refreshToken)
    }

    /**
     * 전달받은 RefreshToken의 유효성을 검증하여 새로운 AccessToken을 재발급합니다.
     * 토큰 만료, 불일치 등의 예외 상황을 세분화하여 처리합니다.
     */
    fun reissue(refreshToken: String): String {
        // 1. 리프레시 토큰 자체의 기본 만료 및 위변조 검증
        if (!jwtTokenProvider.validateToken(refreshToken)) {
            throw TokenReissueException("REFRESH_TOKEN_EXPIRED", "리프레시 토큰이 만료되었습니다. 다시 로그인하십시오.")
        }

        // 2. 토큰 내부 Payload에서 주체(이메일) 추출
        val email = jwtTokenProvider.getUserEmail(refreshToken)
        
        // 3. 추출한 이메일 기반으로 실제 유저 데이터 획득 (엘비스 연산자 적용)
        val user = userRepository.findByEmail(email)
            ?: throw TokenReissueException("REFRESH_TOKEN_USER_NOT_FOUND", "인증 대상 사용자를 찾을 수 없습니다. 다시 로그인하십시오.")

        // 4. 데이터베이스에 저장된 토큰 값과 클라이언트가 보낸 토큰 값 대조
        if (user.refreshToken != refreshToken) {
            throw TokenReissueException("REFRESH_TOKEN_MISMATCH", "유효하지 않은 인증 시도입니다. 다시 로그인하십시오.")
        }

        // 5. 검증 완료 시 새로운 AccessToken만 단독 발행하여 전송
        return jwtTokenProvider.createAccessToken(user.email, user.role.name)
    }

    /**
     * 사용자의 비밀번호를 강제로 재설정합니다.
     * 이메일 인증 코드가 성공적으로 확인된 이후에 호출되어야 안전합니다.
     */
    @Transactional
    fun resetPassword(email: String, newPassword: String) {
        val user = userRepository.findByEmail(email)
            ?: throw EntityNotFoundException("해당 이메일의 사용자를 찾을 수 없습니다.")
        
        user.updatePassword(passwordEncoder.encode(newPassword) ?: throw IllegalStateException("비밀번호 인코딩에 실패했습니다."))
    }

    /**
     * 사용자의 로그아웃을 처리합니다.
     * DB의 리프레시 토큰을 무효화하여 토큰 도용 위험을 차단합니다.
     */
    fun logout(email: String) {
        val user = userRepository.findByEmail(email)
            ?: throw EntityNotFoundException("사용자를 찾을 수 없습니다.")
        user.updateRefreshToken(null)
    }

    /**
     * 닉네임을 통해 가입된 사용자의 이메일 아이디를 추적합니다.
     * 탈퇴 회원 조건은 검색 대상에서 엄격히 배제합니다.
     */
    @Transactional(readOnly = true)
    fun findId(nickname: String): String {
        val user = userRepository.findByNicknameAndRoleNot(nickname, Role.WITHDRAWN)
            ?: throw EntityNotFoundException("해당 닉네임으로 등록된 이메일이 없습니다.")
            
        return user.email
    }

    /**
     * 활성화된 회원을 소프트 딜리트(Soft Delete) 방식으로 탈퇴시킵니다.
     * 데이터 정밀 분석 보존을 위해 상태값만 WITHDRAWN으로 변환합니다.
     */
    fun withdraw(email: String) {
        val user = userRepository.findByEmailAndRoleNot(email, Role.WITHDRAWN)
            ?: throw EntityNotFoundException("활성 사용자를 찾을 수 없습니다.")
        user.withdraw()
    }
}